### ISSUE TITLE
Failure of uniqueness matcher due to mismatched message is a bit wordy

### ISSUE THREAD

**Issue by Drenmi** [thoughtbot/shoulda-matchers#1115]
This seems like an unnecessary amount of text, and readily hides the problem, which is trivial. 🙂 

**Given:**

```ruby
it { is_expected.to validate_uniqueness_of(:user_id).scoped_to(:group_id).with_message("is already a member of this group") }
```

**Expected behaviour:**

Illustrative:

```
Expected GroupsUser uniqueness validation on :user_id scoped to :group_id to fail with message:

  - is already a member of this group

instead got:

  - is already a member of this group.
```

**Actual behaviour:**

```
GroupsUser did not properly validate that :user_id is
       case-sensitively unique within the scope of :group_id, producing a
       custom validation error on failure.
         After taking the given GroupsUser, setting its :user_id to
         ‹"xxxx"› (read back as ‹0›), and saving it as the existing record,
         then making a new GroupsUser and setting its :user_id to a
         different value, ‹0› and its :group_id to a different value, ‹nil›,
         the matcher expected the new GroupsUser to be invalid and
         to produce the validation error "is already a member of this group" on
         :user_id. The record was indeed invalid, but it produced these
         validation errors instead:
     
         * user_id: ["is already a member of this group."]
```

**Comment by mcmire** [comment]
So a bit of background behind this wall of text. We have had various situations where people were trying to use the uniqueness matcher and were very confused as to why their test failed. It is obvious based on these interactions that they generally do not understand how the uniqueness matcher works, so the messaging here is intended to prove that the matcher is not magical and to pull back the curtain to give people extra information when this sort of thing happens.

Of course, I do understand your point about the verbosity. I am wondering now if perhaps we should not have shown this level of detail by default, and instead provided people with some sort of debugging option that could have flipped on at will whenever they were confused about why their test was failing. I use the past tense because, realistically, it is not a trivial thing to change, and based on the status of the project right now, I don't know if this will ever get addressed (it would also complicate the messaging logic, which is so crazy, it just needs to be rewritten at this point). I guess I can leave this open, but I just wanted you to be aware of that.

**Comment by mcmire** [comment]
Okay, cool, I'll close this, then. I definitely appreciate the feedback though!

**Comment by connorshea** [comment]
I would like to echo the sentiment of Dremni's most recent comment. I just spent 15 minutes trying to figure out the source of my test failing, when it was because of a typo ('is aleady in your library' where it was expecting 'is already in your library').

It'd be really nice if it properly diff'd it when the only problem is the message being different.

**Comment by mcmire** [comment]
@connorshea Yeah, I agree in theory that would be nice. The thing about this is that there can be multiple messages on a record, not just one. The matcher doesn't know which one corresponds to the one you were trying to add, so it just prints out all of the messages. We might be able to do some kind of Levenshtein matching to figure that out, and then show some kind of "did you mean" message. Perhaps that would help you get to the root of the problem faster?

**Comment by connorshea** [comment]
Yeah, that’d work as well :)
